flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.util;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.runtime.memory.MemoryAllocationException;
+import org.apache.flink.runtime.memory.MemoryManager;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@link MemorySegmentPool} that lazy allocate memory pages from {@link MemoryManager}.
+ */
+public class LazyMemorySegmentPool implements MemorySegmentPool, Closeable {
+
+	private static final long PER_REQUEST_MEMORY_SIZE = 16 * 1024 * 1024;
+
+	private final Object owner;
+	private final MemoryManager memoryManager;
+	private final boolean freeMore;
+	private final ArrayList<MemorySegment> cachePages;
+	private final int maxPages;
+	private final int perRequestPages;
+
+	private int pageUsage;
+
+	public LazyMemorySegmentPool(Object owner, MemoryManager memoryManager, int maxPages) {
+		this(owner, memoryManager, maxPages, false);
+	}
+
+	public LazyMemorySegmentPool(Object owner, MemoryManager memoryManager, int maxPages, boolean freeMore) {
+		this.owner = owner;
+		this.memoryManager = memoryManager;
+		this.freeMore = freeMore;
+		this.cachePages = new ArrayList<>();
+		this.maxPages = maxPages;
+		this.pageUsage = 0;
+		this.perRequestPages = Math.max(1, (int) (PER_REQUEST_MEMORY_SIZE / memoryManager.getPageSize()));
+	}
+
+	@Override
+	public int pageSize() {
+		return this.memoryManager.getPageSize();
+	}
+
+	@Override
+	public void returnAll(List<MemorySegment> memory) {
+		this.pageUsage -= memory.size();
+		if (this.pageUsage < 0) {
+			throw new RuntimeException("Return too more memories.");
+		}
+		this.cachePages.addAll(memory);
+
+		if (this.freeMore && this.cachePages.size() >= this.perRequestPages) {
+			this.memoryManager.release(this.cachePages);
+		}
+	}
+
+	public void returnPage(MemorySegment segment) {
+		returnAll(Collections.singletonList(segment));
+	}
+
+	@Override
+	public MemorySegment nextSegment() {
+		int freePages = freePages();
+		if (freePages == 0) {
+			return null;
+		}
+
+		if (this.cachePages.isEmpty()) {
+			int numPages = Math.min(freePages, this.perRequestPages);
+			try {
+				this.memoryManager.allocatePages(owner, this.cachePages, numPages);
+			} catch (MemoryAllocationException e) {
+				throw new RuntimeException(e);
+			}
+		}
+		this.pageUsage++;
+		return this.cachePages.remove(this.cachePages.size() - 1);
+	}
+
+	@Override
+	public int freePages() {
+		return this.maxPages - this.pageUsage;
+	}
+
+	@Override
+	public void close() {
+		if (this.pageUsage != 0) {
+			throw new RuntimeException(
+					"Should return all used memory before clean, page used: " + pageUsage);
+		}
+		cleanCache();
+	}
+
+	public void cleanCache() {
+		this.memoryManager.release(this.cachePages);
```

Comment:
Shall we check if `this.cachePages.size() >= this.perRequestPages`?

Reply:
No, this is a manual release of excess memory. This frequency will not be too high.